File: Java/Arrays/Medium/Sort an array of 0s, 1s and 2s/Solution.java
import java.util.Scanner;
/*  
 * For the Counting Sort
 * TC = O(2N)
 * SC = O(1)
 * 
 * For Dutch National Flag Algorithm
 * TC = O(n) - Single Pass Algo
 * SC = O(1)
 */
public class Solution {
    private static void printArray(int[] arr){
        int n = arr.length;
        for (int i = 0; i<n; i++){
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
    // Dutch National Flag Algorithm (DNFA)
    private static void swap(int[] arr, int index1, int index2) {
    int temp = arr[index1];
    arr[index1] = arr[index2];
    arr[index2] = temp;
}
    private static void sortColors(int[] nums) {
        int n = nums.length;
        int low = 0, mid = 0, high = n - 1;
        while (mid <= high) {
            switch (nums[mid]) {
                case 0:
                    swap(nums, low, mid);
                    low++;
                    mid++;
                    break;
                case 1:
                    mid++;
                    break;
                case 2:
                    swap(nums, mid, high);
                    high--;
                    break;
            }
        }
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();   
        while (t >= 1){ 
            t--;
            int n = sc.nextInt();
            int[] arr = new int[n];
            for (int i = 0; i < n; i++){
                arr[i] = sc.nextInt();
            }
            sortColors(arr);
            printArray(arr);
        }
        sc.close();
    }
}

File: Python/Arrays/Medium/Sort an array of 0s, 1s and 2s/Solution.py
''' 
    For the Counting Sort
    TC = O(2N)
    SC = O(1)

    For Dutch National Flag Algorithm
    TC = O(n) - Single Pass Algo
    SC = O(1)
'''
from typing import List

def sortColors(nums: List[int]) -> None:
    low, mid = 0, 0
    high = len(nums) - 1
    while mid <= high:
        match nums[mid]:
            case 0:
                nums[low], nums[mid] = nums[mid], nums[low]
                low += 1
                mid += 1
            case 1:
                mid += 1
            case 2:
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1
    
# Main execution block: Handles input/output for testing
t = int(input()) # Read number of test cases
while t > 0:
    t -= 1
    # Read elements for arr
    arr = list(map(int, input().split()))
    sortColors(arr)
    print(arr)
